File: agent/library/src/main/java/org/matilda/messages/MessageHandlerFactory.java
package org.matilda.messages;

import org.matilda.commands.CommandMessageHandler;
import org.matilda.messages.handlers.MessageDispatcher;
import org.matilda.messages.handlers.MessageHandler;
import org.matilda.messages.protobuf.MessageType;

import javax.inject.Inject;
import java.util.concurrent.ExecutorService;

public class MessageHandlerFactory {
    @Inject
    ExecutorService mExecutorService;

    @Inject
    MessageHandlerRegistry mMessageHandlerRegistry;

    @Inject
    CommandMessageHandler mCommandMessageHandler;

    @Inject
    public MessageHandlerFactory() {}

    public MessageHandler create() {
        mMessageHandlerRegistry.registerHandler(MessageType.COMMAND.getNumber(), mCommandMessageHandler);

        return new MessageDispatcher(mExecutorService, mMessageHandlerRegistry);
    }
}

File: agent/library/src/main/java/org/matilda/messages/MessageHandlerFactory.kt
package org.matilda.messages

import kotlinx.coroutines.CoroutineScope
import me.tatarka.inject.annotations.Inject
import org.matilda.commands.CommandMessageHandler
import org.matilda.messages.handlers.MessageDispatcher
import org.matilda.messages.handlers.MessageHandler
import org.matilda.messages.protobuf.MessageType

@Inject
class MessageHandlerFactory(private val mCoroutineScope: CoroutineScope,
                            private val mMessageHandlerRegistry: MessageHandlerRegistry,
                            private val mCommandMessageHandler: CommandMessageHandler) {

    fun create(): MessageHandler {
        mMessageHandlerRegistry.registerHandler(MessageType.COMMAND.value, mCommandMessageHandler)
        return MessageDispatcher(mCoroutineScope, mMessageHandlerRegistry)
    }
}
